### ISSUE TITLE
helm dependency build or helm lint fail with local dependencies with a pre-release version

### ISSUE THREAD

**Issue by gernoteger** [helm/helm#11855]
We have a chart with a local, verbatim subchart in the` /charts` directory.
The subchart has a version that contains a SemVer pre-release value

Starting with helm 3.4.0, we have this issue:

- if the outer chart doesn't reference the subchart, `helm lint` fails: `chart metadata is missing these dependencies: bar`
- if the outer chart references the subchart, `helm dependency build` fails: `dependency bar at version 0.1.0-SNAPSHOT does not satisfy the constraint * `

when the subchart's version doesn't contain the pre-release version part, `helm dependency build` works  nicely.

In order to extract and reproduce the issue,I created a simple example: https://github.com/gernoteger/helm-compatibility-tests
In our real life applications, the version is created by our CI/CD Pipeline.

From the helm documentation I assume that a version with a pre-release tag is perfectly valid, therefore I think this is a bug in the implentation of either lint, or mnore likely, the helm depency SemVer resolution.

**Comment by gjenkins8** [comment]
The purpose of the `helm dependency build` command is to:

```
Build out the charts/ directory from the Chart.lock file.

Build is used to reconstruct a chart's dependencies to the state specified in
the lock file. This will not re-negotiate dependencies, as 'helm dependency update'
does.

If no lock file is found, 'helm dependency build' will mirror the behavior
of 'helm dependency update'.
```

But you already have the `chart/` dir populated with `bar/` (chart and directory). Also note the last paragraph in the help. Since you don't have a `Chart.lock`, you are really doing a `helm dependency update`. But overall, for your (simple) example, `helm dependency build|update` is redundant if the `chart/` directory is already populated I think (perhaps you have more subcharts in your actual use?)

Finally, semver treats pre-release versions specially. `version: "*"` means match highest major version. `version: "*-0"` means to include pre-release versions. The reason is valid: you want to opt-in to matching pre-release versions. Otherwise e.g. if today you match version `2.3.4`, then tomorrow you match `3.0.0-superbuggyprereleasealphadontuseinforeal`, that would be annoying/unexpected for users who value stability. Indeed, if you chart foo's `Chart.yaml` to accept prerelease versions, `helm dependency build|update` work for your example.

As for what changed (it sounds like you were/are using a very old version of helm, older than 3.4.0). I'm unsure. If I had to guess, helm or the semver library it uses was changed to not match prerelease versions.
